File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibraryRulesheetSupport.kt
package com.pillyliu.pinprofandroid.library

internal data class ReferenceLink(
    val label: String,
    val url: String? = null,
) {
    val embeddedRulesheetSource: RulesheetRemoteSource?
        get() {
            val destination = resolveLibraryUrl(destinationUrl) ?: return null
            val normalized = destination.lowercase()
            if (normalized.contains("pinballnews.com")) return null
            return when {
                normalized.contains("tiltforums.com") -> RulesheetRemoteSource.TiltForums(destination)
                normalized.contains("rules.silverballmania.com") ||
                    normalized.contains("silverballmania.com") ||
                    normalized.contains("flippers.be") ||
                    label.lowercase().contains("(bob)") -> RulesheetRemoteSource.BobsGuide(destination)
                normalized.contains("pinballprimer.github.io") ||
                    normalized.contains("pinballprimer.com") ||
                    label.lowercase().contains("(pp)") -> RulesheetRemoteSource.PinballPrimer(destination)
                normalized.contains("replayfoundation.org") ||
                    normalized.contains("pinball.org") ||
                    label.lowercase().contains("(papa)") -> RulesheetRemoteSource.Papa(destination)
                else -> null
            }
        }

    val destinationUrl: String?
        get() = url?.trim()?.ifBlank { null }
}

internal enum class RulesheetSourceKind(val rank: Int, val shortLabel: String) {
    LOCAL(0, "Local"),
    TF(1, "TF"),
    PROF(2, "PinProf"),
    BOB(3, "Bob"),
    PAPA(4, "PAPA"),
    PP(5, "PP"),
    OPDB(6, "OPDB"),
    OTHER(7, "Other"),
}

internal val ReferenceLink.rulesheetSourceKind: RulesheetSourceKind
    get() {
        val normalizedLabel = label.lowercase()
        val resolved = resolveLibraryUrl(destinationUrl)
        return when {
            isPinProfRulesheetUrl(resolved) || "(prof)" in normalizedLabel -> RulesheetSourceKind.PROF
            resolved?.contains("tiltforums.com", ignoreCase = true) == true || "(tf)" in normalizedLabel -> RulesheetSourceKind.TF
            resolved?.contains("pinballprimer.github.io", ignoreCase = true) == true ||
                resolved?.contains("pinballprimer.com", ignoreCase = true) == true ||
                "(pp)" in normalizedLabel -> RulesheetSourceKind.PP
            resolved?.contains("replayfoundation.org", ignoreCase = true) == true ||
                resolved?.contains("pinball.org", ignoreCase = true) == true ||
                "(papa)" in normalizedLabel -> RulesheetSourceKind.PAPA
            resolved?.contains("silverballmania.com", ignoreCase = true) == true ||
                resolved?.contains("flippers.be", ignoreCase = true) == true ||
                "(bob)" in normalizedLabel -> RulesheetSourceKind.BOB
            "(opdb)" in normalizedLabel -> RulesheetSourceKind.OPDB
            "(local)" in normalizedLabel || "(source)" in normalizedLabel -> RulesheetSourceKind.LOCAL
            resolved == null && embeddedRulesheetSource == null -> RulesheetSourceKind.LOCAL
            else -> RulesheetSourceKind.OTHER
        }
    }

internal val ReferenceLink.shortRulesheetTitle: String
    get() = rulesheetSourceKind.shortLabel

internal data class Video(val kind: String?, val label: String?, val url: String?)

internal data class PlayableVideo(val id: String, val label: String) {
    val watchUrl: String
        get() = "https://www.youtube.com/watch?v=$id"

    val thumbnailUrl: String
        get() = "https://i.ytimg.com/vi/$id/hqdefault.jpg"
}

internal data class YouTubeVideoMetadata(val title: String, val channelName: String?)

sealed interface RulesheetRemoteSource {
    val url: String
    val sourceName: String
    val originalLinkLabel: String
    val detailsText: String

    data class TiltForums(override val url: String) : RulesheetRemoteSource {
        override val sourceName: String = "Tilt Forums community rulesheet"
        override val originalLinkLabel: String = "Original thread"
        override val detailsText: String = "License/source terms remain with Tilt Forums and the original authors."
    }

    data class BobsGuide(override val url: String) : RulesheetRemoteSource {
        override val sourceName: String = "Silverball Rules (Bob Matthews source)"
        override val originalLinkLabel: String = "Original page"
        override val detailsText: String = "Preserve source attribution and any author/site rights notes from the original page."
    }

    data class PinballPrimer(override val url: String) : RulesheetRemoteSource {
        override val sourceName: String = "Pinball Primer"
        override val originalLinkLabel: String = "Original page"
        override val detailsText: String = "Preserve source attribution and any author/site rights notes from the original page."
    }

    data class Papa(override val url: String) : RulesheetRemoteSource {
        override val sourceName: String = "PAPA / pinball.org rulesheet archive"
        override val originalLinkLabel: String = "Original page"
        override val detailsText: String = "Preserve source attribution and any author/site rights notes from the original page."
    }
}

File: Pinball App 2/Pinball App 2/library/LibraryRulesheetSupport.swift
import Foundation

enum LibraryRulesheetSourceKind: Int {
    case local = 0
    case tf = 1
    case prof = 2
    case bob = 3
    case papa = 4
    case pp = 5
    case opdb = 6
    case other = 7

    nonisolated var shortTitle: String {
        switch self {
        case .local:
            return "Local"
        case .prof:
            return "PinProf"
        case .bob:
            return "Bob"
        case .papa:
            return "PAPA"
        case .pp:
            return "PP"
        case .tf:
            return "TF"
        case .opdb:
            return "OPDB"
        case .other:
            return "Other"
        }
    }
}

nonisolated func libraryIsPinProfRulesheetURL(_ url: URL?) -> Bool {
    guard let url,
          libraryIsPinProfHost(url.host) else {
        return false
    }
    return url.path.hasPrefix("/pinball/rulesheets/")
}

nonisolated func normalizedRulesheetMarkdownPath(_ pathOrURL: String?) -> String? {
    guard let raw = pathOrURL?.trimmingCharacters(in: .whitespacesAndNewlines),
          !raw.isEmpty,
          let resolved = libraryResolveURL(pathOrURL: raw)?.path.lowercased(),
          !resolved.isEmpty else {
        return nil
    }
    return resolved
}

nonisolated func libraryIsLikelyPinProfMarkdownRulesheetURL(_ url: URL?) -> Bool {
    guard let raw = url?.absoluteString.trimmingCharacters(in: .whitespacesAndNewlines),
          !raw.isEmpty else {
        return false
    }
    let normalizedRaw = raw.lowercased()
    if normalizedRaw.hasSuffix("-rulesheet.md") ||
        normalizedRaw.contains("/pinball/rulesheets/") ||
        (normalizedRaw.contains("/rules/") && normalizedRaw.contains("source=local")) {
        return true
    }
    guard let resolvedPath = normalizedRulesheetMarkdownPath(raw) else {
        return false
    }
    return resolvedPath.hasPrefix("/pinball/rulesheets/") ||
        resolvedPath.hasSuffix("-rulesheet.md") ||
        (resolvedPath.hasPrefix("/rules/") && normalizedRaw.contains("source=local"))
}

extension PinballGame.ReferenceLink {
    nonisolated var rulesheetSourceKind: LibraryRulesheetSourceKind {
        let normalizedLabel = label.lowercased()
        let resolvedURL = libraryResolveURL(pathOrURL: url)

        if libraryIsPinProfRulesheetURL(resolvedURL) || normalizedLabel.contains("(prof)") {
            return .prof
        }
        if resolvedURL?.host?.lowercased().contains("tiltforums.com") == true || normalizedLabel.contains("(tf)") {
            return .tf
        }
        if resolvedURL?.host?.lowercased().contains("pinballprimer.github.io") == true
            || resolvedURL?.host?.lowercased().contains("pinballprimer.com") == true
            || normalizedLabel.contains("(pp)") {
            return .pp
        }
        if resolvedURL?.host?.lowercased().contains("pinball.org") == true
            || resolvedURL?.host?.lowercased().contains("replayfoundation.org") == true
            || normalizedLabel.contains("(papa)") {
            return .papa
        }
        if resolvedURL?.host?.lowercased().contains("silverballmania.com") == true
            || resolvedURL?.host?.lowercased().contains("flippers.be") == true
            || normalizedLabel.contains("(bob)") {
            return .bob
        }
        if normalizedLabel.contains("(opdb)") {
            return .opdb
        }
        if normalizedLabel.contains("(local)") || normalizedLabel.contains("(source)") {
            return .local
        }
        if resolvedURL == nil && embeddedRulesheetSource == nil {
            return .local
        }
        return .other
    }

    nonisolated var shortRulesheetTitle: String {
        rulesheetSourceKind.shortTitle
    }

    nonisolated var rulesheetSortKey: (Int, String, String) {
        (
            rulesheetSourceKind.rawValue,
            label.lowercased(),
            (libraryResolveURL(pathOrURL: url)?.absoluteString ?? url).lowercased()
        )
    }
}
